# **Recommender System Implementation in Jupyter Notebook**

A **Recommender System** is a type of information filtering system that predicts user preferences and suggests relevant items (movies, products, music, etc.). They are widely used by platforms like Netflix, Amazon, and Spotify.

---

## **Types of Recommender Systems**
There are **three main types** of recommender systems:

1. **Collaborative Filtering**  
   - Recommends items based on user behavior and preferences of similar users.  
   - **Example:** "Users who liked this movie also liked…"  
   - **Types:**  
     - **User-User CF:** Finds similar users and recommends items they liked.  
     - **Item-Item CF:** Finds similar items based on user interactions.  

2. **Content-Based Filtering**  
   - Recommends items similar to those a user has liked in the past, based on item features.  
   - **Example:** If a user likes action movies, recommend other action movies.  

3. **Hybrid Recommender System**  
   - Combines **collaborative** and **content-based** approaches for better accuracy.  
   - **Example:** Netflix uses both viewing history (collaborative) and movie genres (content-based).  



# **Location, Price & Facilities-Based Hybrid Recommender System**  
***(With Custom Weighted Recommendations)***  

Here, we are building a **hybrid recommender system** that combines:  
1. **Location-based recommendations** (proximity to key landmarks)  
2. **Price-based recommendations** (affordability & value)  
3. **Facilities-based recommendations** (top amenities)  

Each system generates scores independently, and we **assign weights** to combine them into a final recommendation.



In [ ]:
import numpy as np
import pandas as pd
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
df = pd.read_csv('appartments.csv').drop(22)

In [ ]:
df.head()

PropertyName  \
0           Smartworld One DXP   
1                    M3M Crown   
2  Adani Brahma Samsara Vilasa   
3                   Sobha City   
4     Signature Global City 93   

                                     PropertySubName  \
0       2, 3, 4 BHK Apartment in Sector 113, Gurgaon   
1          3, 4 BHK Apartment in Sector 111, Gurgaon   
2  Land, 3, 4 BHK Independent Floor in Sector 63,...   
3       2, 3, 4 BHK Apartment in Sector 108, Gurgaon   
4    2, 3 BHK Independent Floor in Sector 93 Gurgaon   

                                     NearbyLocations  \
0  ['Bajghera Road', 'Palam Vihar Halt', 'DPSG Pa...   
1  ['DPSG Palam Vihar Gurugram', 'The NorthCap Un...   
2  ['AIPL Business Club Sector 62', 'Heritage Xpe...   
3  ['The Shikshiyan School', 'WTC Plaza', 'Luxus ...   
4  ['Pranavananda Int. School', 'DLF Site central...   

                                  LocationAdvantages  \
0  {'Bajghera Road': '800 Meter', 'Palam Vihar Ha...   
1  {'DPSG Palam Vihar Gurugram': '1.4 Km', 'The N...   
2  {'AIPL Business Club Sector 62': '2.7 Km', 'He...   
3  {'The Shikshiyan School': '2.9 KM', 'WTC Plaza...   
4  {'Pranavananda Int. School': '450 m', 'DLF Sit...   

                                                Link  \
0  https://www.99acres.com/smartworld-one-dxp-sec...   
1  https://www.99acres.com/m3m-crown-sector-111-g...   
2  https://www.99acres.com/adani-brahma-samsara-v...   
3  https://www.99acres.com/sobha-city-sector-108-...   
4  https://www.99acres.com/signature-global-city-...   

                                        PriceDetails  \
0  {'2 BHK': {'building_type': 'Apartment', 'area...   
1  {'3 BHK': {'building_type': 'Apartment', 'area...   
2  {'3 BHK': {'building_type': 'Independent Floor...   
3  {'2 BHK': {'building_type': 'Apartment', 'area...   
4  {'2 BHK': {'building_type': 'Independent Floor...   

                                       TopFacilities  
0  ['Swimming Pool', 'Salon', 'Restaurant', 'Spa'...  
1  ['Bowling Alley', 'Mini Theatre', 'Manicured G...  
2  ['Terrace Garden', 'Gazebo', 'Fountain', 'Amph...  
3  ['Swimming Pool', 'Volley Ball Court', 'Aerobi...  
4  ['Mini Theatre', 'Doctor on Call', 'Concierge ...

Here’s a concise summary of each feature in your real estate dataset:

---

### **1. PropertyName**  
- **Summary**: Name of the property/residential project.  
- **Example**: `Smartworld One DXP`, `M3M Crown`.

### **2. PropertySubName**  
- **Summary**: Sub-description of the property (e.g., BHK configuration + location).  
- **Example**: `2, 3, 4 BHK Apartment in Sector 113, Gurgaon`.

### **3. NearbyLocations**  
- **Summary**: List of nearby landmarks/points of interest.  
- **Example**: `['Bajghera Road', 'Palam Vihar Halt', 'DPSG School']`.

### **4. LocationAdvantages**  
- **Summary**: Dictionary of nearby locations with their distances.  
- **Example**: `{'Bajghera Road': '800 Meter', 'Palam Vihar Halt': '1.2 Km'}`.

### **5. Link**  
- **Summary**: URL to the property’s listing page (e.g., 99acres.com).  
- **Example**: `https://www.99acres.com/smartworld-one-dxp...`.

### **6. PriceDetails**  
- **Summary**: Nested dictionary of BHK configurations with details like:  
  - `building_type` (Apartment/Independent Floor).  
  - `area` (sq.ft).  
  - `price` (in ₹).  
- **Example**:  
  ```python
  {'2 BHK': {'building_type': 'Apartment', 'area': '1200 sq.ft', 'price': '1.2 Cr'}}
  ```

### **7. TopFacilities**  
- **Summary**: List of amenities offered by the property.  
- **Example**: `['Swimming Pool', 'Salon', 'Restaurant', 'Spa']`.

---



In [ ]:
df.columns

Index(['PropertyName', 'PropertySubName', 'NearbyLocations',
       'LocationAdvantages', 'Link', 'PriceDetails', 'TopFacilities'],
      dtype='object')

In [ ]:
df.iloc[0].PropertySubName

'2, 3, 4 BHK Apartment in Sector 113, Gurgaon'

In [ ]:
df.iloc[0].NearbyLocations

"['Bajghera Road', 'Palam Vihar Halt', 'DPSG Palam Vihar', 'Park Hospital', 'Gurgaon Railway Station']"

In [ ]:
df.iloc[0].LocationAdvantages

"{'Bajghera Road': '800 Meter', 'Palam Vihar Halt': '2.5 KM', 'DPSG Palam Vihar': '3.1 KM', 'Park Hospital': '3.1 KM', 'Gurgaon Railway Station': '4.9 KM', 'The NorthCap University': '5.4 KM', 'Dwarka Expy': '1.2 KM', 'Hyatt Place Gurgaon Udyog Vihar': '7.7 KM', 'Dwarka Sector 21, Metro Station': '7.2 KM', 'Pacific D21 Mall': '7.4 KM', 'Indira Gandhi International Airport': '14.7 KM', 'Hamoni Golf Camp': '6.2 KM', 'Fun N Food Waterpark': '8.8 KM', 'Accenture DDC5': '9 KM'}"

In [ ]:
df.iloc[0].Link

'https://www.99acres.com/smartworld-one-dxp-sector-113-gurgaon-npxid-r400415'

In [ ]:
df.iloc[0].PriceDetails

"{'2 BHK': {'building_type': 'Apartment', 'area_type': 'Carpet Area', 'area': '1,370 sq.ft.', 'price-range': '₹ 2 - 2.4 Cr'}, '3 BHK': {'building_type': 'Apartment', 'area_type': 'Carpet Area', 'area': '1,850 - 2,050 sq.ft.', 'price-range': '₹ 2.25 - 3.59 Cr'}, '4 BHK': {'building_type': 'Apartment', 'area_type': 'Carpet Area', 'area': '2,600 sq.ft.', 'price-range': '₹ 3.24 - 4.56 Cr'}}"

In [ ]:
df.iloc[0].TopFacilities

"['Swimming Pool', 'Salon', 'Restaurant', 'Spa', 'Cafeteria', 'Sun Deck', '24x7 Security', 'Club House', 'Gated Community']"

### Observations:
* NearbyLocations is subset of LocationAdvantages
* PropertySubName:  we are skipping this information(LocationAdvantages consist this info)
* Link: useful when displaying
* PriceDetails: useful
* TopFacilities: useful

### **Facility-Based Recommendation System Using Vector Similarity**

To enable intelligent facility-based recommendations, we transform each property's amenities into a structured numerical representation (feature vector) and compute similarity scores between properties using **cosine similarity**. This approach identifies properties with the most comparable facilities, allowing us to recommend the top 5 most relevant options.  

#### **Methodology**  
1. **Feature Vectorization**  
   - Each facility (e.g., "wifi," "pool," "gym") is assigned a binary or weighted value (1/0 or custom weights) to create a numerical vector per property.  
   - Example:  
     - Property A: `[1, 1, 0]` → "wifi, pool"  
     - Property B: `[1, 0, 1]` → "wifi, gym"  

2. **Similarity Computation**  
   - **Cosine similarity** measures the angular distance between vectors, ranging from 0 (no overlap) to 1 (identical facilities).  
   - Formula:  
     \[
     \text{Similarity} = \frac{A \cdot B}{\|A\| \|B\|}
     \]
   - Higher scores indicate more aligned amenities.  

3. **Top-5 Recommendations**  
   - For a target property, compute similarity scores against all others.  
   - Rank properties by descending similarity and select the top 5.  

#### **Advantages**  
- **Scalability**: Handles large inventories efficiently.  
- **Context-Aware**: Prioritizes properties with matching/upgraded amenities.  
- **Dynamic Weights**: Customizable for premium facilities (e.g., "pool" weighted higher than "parking").  



## TopFacilities

In [ ]:
df[['PropertyName','TopFacilities']]

PropertyName  \
0             Smartworld One DXP   
1                      M3M Crown   
2    Adani Brahma Samsara Vilasa   
3                     Sobha City   
4       Signature Global City 93   
..                           ...   
242         DLF Princeton Estate   
243        Pyramid Urban Homes 2   
244          Satya The Hermitage   
245                  BPTP Spacio   
246             SS The Coralwood   

                                         TopFacilities  
0    ['Swimming Pool', 'Salon', 'Restaurant', 'Spa'...  
1    ['Bowling Alley', 'Mini Theatre', 'Manicured G...  
2    ['Terrace Garden', 'Gazebo', 'Fountain', 'Amph...  
3    ['Swimming Pool', 'Volley Ball Court', 'Aerobi...  
4    ['Mini Theatre', 'Doctor on Call', 'Concierge ...  
..                                                 ...  
242  ['Swimming Pool', 'Medical Centre', 'Laundry',...  
243  ['Shopping Centre', 'Community Hall', '24x7 Se...  
244  ['Bus Shelter', 'Swimming Pool', 'Business Lou...  
245  ['Swimming Pool', 'Card Room', 'Piped Gas', 'P...  
246  ['Swimming Pool', 'School', 'Solar Lighting', ...  

[246 rows x 2 columns]

In [ ]:
df[['PropertyName','TopFacilities']]['TopFacilities'][0]

"['Swimming Pool', 'Salon', 'Restaurant', 'Spa', 'Cafeteria', 'Sun Deck', '24x7 Security', 'Club House', 'Gated Community']"

In [ ]:
def extract_list(s):
    return re.findall(r"'(.*?)'", s)

df['TopFacilities'] = df['TopFacilities'].apply(extract_list)

In [ ]:
df.head()

PropertyName  \
0           Smartworld One DXP   
1                    M3M Crown   
2  Adani Brahma Samsara Vilasa   
3                   Sobha City   
4     Signature Global City 93   

                                     PropertySubName  \
0       2, 3, 4 BHK Apartment in Sector 113, Gurgaon   
1          3, 4 BHK Apartment in Sector 111, Gurgaon   
2  Land, 3, 4 BHK Independent Floor in Sector 63,...   
3       2, 3, 4 BHK Apartment in Sector 108, Gurgaon   
4    2, 3 BHK Independent Floor in Sector 93 Gurgaon   

                                     NearbyLocations  \
0  ['Bajghera Road', 'Palam Vihar Halt', 'DPSG Pa...   
1  ['DPSG Palam Vihar Gurugram', 'The NorthCap Un...   
2  ['AIPL Business Club Sector 62', 'Heritage Xpe...   
3  ['The Shikshiyan School', 'WTC Plaza', 'Luxus ...   
4  ['Pranavananda Int. School', 'DLF Site central...   

                                  LocationAdvantages  \
0  {'Bajghera Road': '800 Meter', 'Palam Vihar Ha...   
1  {'DPSG Palam Vihar Gurugram': '1.4 Km', 'The N...   
2  {'AIPL Business Club Sector 62': '2.7 Km', 'He...   
3  {'The Shikshiyan School': '2.9 KM', 'WTC Plaza...   
4  {'Pranavananda Int. School': '450 m', 'DLF Sit...   

                                                Link  \
0  https://www.99acres.com/smartworld-one-dxp-sec...   
1  https://www.99acres.com/m3m-crown-sector-111-g...   
2  https://www.99acres.com/adani-brahma-samsara-v...   
3  https://www.99acres.com/sobha-city-sector-108-...   
4  https://www.99acres.com/signature-global-city-...   

                                        PriceDetails  \
0  {'2 BHK': {'building_type': 'Apartment', 'area...   
1  {'3 BHK': {'building_type': 'Apartment', 'area...   
2  {'3 BHK': {'building_type': 'Independent Floor...   
3  {'2 BHK': {'building_type': 'Apartment', 'area...   
4  {'2 BHK': {'building_type': 'Independent Floor...   

                                       TopFacilities  
0  [Swimming Pool, Salon, Restaurant, Spa, Cafete...  
1  [Bowling Alley, Mini Theatre, Manicured Garden...  
2  [Terrace Garden, Gazebo, Fountain, Amphitheatr...  
3  [Swimming Pool, Volley Ball Court, Aerobics Ce...  
4  [Mini Theatre, Doctor on Call, Concierge Servi...

In [ ]:
df['FacilitiesStr'] = df['TopFacilities'].apply(' '.join)

In [ ]:
df['FacilitiesStr'][0]

'Swimming Pool Salon Restaurant Spa Cafeteria Sun Deck 24x7 Security Club House Gated Community'

In [ ]:
tfidf_vectorizer = TfidfVectorizer(stop_words='english', ngram_range=(1, 2))

In [ ]:
tfidf_matrix = tfidf_vectorizer.fit_transform(df['FacilitiesStr'])

In [ ]:
tfidf_matrix.shape # it is a 247 vectors in a 953 dimension spaces

(246, 953)

In [ ]:
 tfidf_matrix.toarray()[0]

array([0.        , 0.        , 0.        , 0.18809342, 0.18809342,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.     

In [ ]:
# calculating cosine_similarity array
cosine_sim1 = cosine_similarity(tfidf_matrix, tfidf_matrix)

### This is 247*247 matrix because it consist of 247 vectors distance with other 247 vectors. each vector represent the property.

In [ ]:
cosine_sim1.shape

(246, 246)

In [ ]:
def recommend_properties(property_name, cosine_sim=cosine_sim1):
    # Get the index of the property that matches the name
    idx = df.index[df['PropertyName'] == property_name].tolist()[0]

    # Get the pairwise similarity scores with that property
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the properties based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar properties
    sim_scores = sim_scores[1:6]

    # Get the property indices
    property_indices = [i[0] for i in sim_scores]

    recommendations_df = pd.DataFrame({
        'PropertyName': df['PropertyName'].iloc[property_indices],
        'SimilarityScore': sim_scores
    })

    # Return the top 10 most similar properties
    return recommendations_df

In [ ]:
recommend_properties("DLF The Arbour")

PropertyName            SimilarityScore
64      Ace Palm Floors   (63, 0.4529382062441955)
217         Yashika 104  (216, 0.4199606322926784)
93       JMS The Nation   (92, 0.4166584649363288)
154       India Rashtra   (153, 0.398954234680194)
0    Smartworld One DXP   (0, 0.38885046199432893)

In [ ]:
df.iloc[11]

PropertyName                                             DLF The Arbour
PropertySubName                    4 BHK Apartment in Sector 63 Gurgaon
NearbyLocations       ['WorldMark Gurgaon', 'Capital Cyberscape', 'T...
LocationAdvantages    {'WorldMark Gurgaon': '1.5 KM', 'Capital Cyber...
Link                  https://www.99acres.com/dlf-the-arbour-sector-...
PriceDetails          {'4 BHK': {'building_type': 'Apartment', 'area...
TopFacilities         [Swimming Pool, Community Hall, Yoga/Meditatio...
FacilitiesStr         Swimming Pool Community Hall Yoga/Meditation A...
Name: 11, dtype: object

## PriceDetails

In [ ]:
df[['PropertyName','PriceDetails']]['PriceDetails'][1]

"{'3 BHK': {'building_type': 'Apartment', 'area_type': 'Super Built-up Area', 'area': '1,605 - 2,170 sq.ft.', 'price-range': '₹ 2.2 - 3.03 Cr'}, '4 BHK': {'building_type': 'Apartment', 'area_type': 'Super Built-up Area', 'area': '2,248 - 2,670 sq.ft.', 'price-range': '₹ 3.08 - 3.73 Cr'}}"

In [ ]:
import pandas as pd
import json

# Load the dataset
df_appartments = pd.read_csv('appartments.csv').drop(22)

# Function to parse and extract the required features from the PriceDetails column
def refined_parse_modified_v2(detail_str):
    try:
        details = json.loads(detail_str.replace("'", "\""))

    except:
        return {}

    extracted = {}
    for bhk, detail in details.items():
        # Extract building type
        extracted[f'building type_{bhk}'] = detail.get('building_type')

        # Parsing area details
        area = detail.get('area', '')
        area_parts = area.split('-')
        if len(area_parts) == 1:
            try:
                value = float(area_parts[0].replace(',', '').replace(' sq.ft.', '').strip())
                extracted[f'area low {bhk}'] = value
                extracted[f'area high {bhk}'] = value
            except:
                extracted[f'area low {bhk}'] = None
                extracted[f'area high {bhk}'] = None
        elif len(area_parts) == 2:
            try:
                extracted[f'area low {bhk}'] = float(area_parts[0].replace(',', '').replace(' sq.ft.', '').strip())
                extracted[f'area high {bhk}'] = float(area_parts[1].replace(',', '').replace(' sq.ft.', '').strip())
            except:
                extracted[f'area low {bhk}'] = None
                extracted[f'area high {bhk}'] = None

        # Parsing price details
        price_range = detail.get('price-range', '')
        price_parts = price_range.split('-')
        if len(price_parts) == 2:
            try:
                extracted[f'price low {bhk}'] = float(price_parts[0].replace('₹', '').replace(' Cr', '').replace(' L', '').strip())
                extracted[f'price high {bhk}'] = float(price_parts[1].replace('₹', '').replace(' Cr', '').replace(' L', '').strip())
                if 'L' in price_parts[0]:
                    extracted[f'price low {bhk}'] /= 100
                if 'L' in price_parts[1]:
                    extracted[f'price high {bhk}'] /= 100
            except:
                extracted[f'price low {bhk}'] = None
                extracted[f'price high {bhk}'] = None

    return extracted
# Apply the refined parsing and generate the new DataFrame structure
data_refined = []

for _, row in df_appartments.iterrows():
    features = refined_parse_modified_v2(row['PriceDetails'])

    # Construct a new row for the transformed dataframe
    new_row = {'PropertyName': row['PropertyName']}

    # Populate the new row with extracted features
    for config in ['1 BHK', '2 BHK', '3 BHK', '4 BHK', '5 BHK', '6 BHK', '1 RK', 'Land']:
        new_row[f'building type_{config}'] = features.get(f'building type_{config}')
        new_row[f'area low {config}'] = features.get(f'area low {config}')
        new_row[f'area high {config}'] = features.get(f'area high {config}')
        new_row[f'price low {config}'] = features.get(f'price low {config}')
        new_row[f'price high {config}'] = features.get(f'price high {config}')

    data_refined.append(new_row)

df_final_refined_v2 = pd.DataFrame(data_refined).set_index('PropertyName')


In [ ]:
df_final_refined_v2.head()

building type_1 BHK  area low 1 BHK  \
PropertyName                                                      
Smartworld One DXP                         None             NaN   
M3M Crown                                  None             NaN   
Adani Brahma Samsara Vilasa                None             NaN   
Sobha City                                 None             NaN   
Signature Global City 93                   None             NaN   

                             area high 1 BHK  price low 1 BHK  \
PropertyName                                                    
Smartworld One DXP                       NaN              NaN   
M3M Crown                                NaN              NaN   
Adani Brahma Samsara Vilasa              NaN              NaN   
Sobha City                               NaN              NaN   
Signature Global City 93                 NaN              NaN   

                             price high 1 BHK building type_2 BHK  \
PropertyName                                                        
Smartworld One DXP                        NaN           Apartment   
M3M Crown                                 NaN                None   
Adani Brahma Samsara Vilasa               NaN                None   
Sobha City                                NaN           Apartment   
Signature Global City 93                  NaN   Independent Floor   

                             area low 2 BHK  area high 2 BHK  price low 2 BHK  \
PropertyName                                                                    
Smartworld One DXP                   1370.0           1370.0           2.0000   
M3M Crown                               NaN              NaN              NaN   
Adani Brahma Samsara Vilasa             NaN              NaN              NaN   
Sobha City                           1381.0           1692.0           1.5500   
Signature Global City 93              981.0           1118.0           0.9301   

                             price high 2 BHK  ... building type_1 RK  \
PropertyName                                   ...                      
Smartworld One DXP                       2.40  ...               None   
M3M Crown                                 NaN  ...               None   
Adani Brahma Samsara Vilasa               NaN  ...               None   
Sobha City                               3.21  ...               None   
Signature Global City 93                 1.06  ...               None   

                             area low 1 RK  area high 1 RK  price low 1 RK  \
PropertyName                                                                 
Smartworld One DXP                     NaN             NaN             NaN   
M3M Crown                              NaN             NaN             NaN   
Adani Brahma Samsara Vilasa            NaN             NaN             NaN   
Sobha City                             NaN             NaN             NaN   
Signature Global City 93               NaN             NaN             NaN   

                             price high 1 RK building type_Land  \
PropertyName                                                      
Smartworld One DXP                       NaN               None   
M3M Crown                                NaN               None   
Adani Brahma Samsara Vilasa              NaN                      
Sobha City                               NaN               None   
Signature Global City 93                 NaN               None   

                             area low Land  area high Land  price low Land  \
PropertyName                                                                 
Smartworld One DXP                     NaN             NaN             NaN   
M3M Crown                              NaN             NaN             NaN   
Adani Brahma Samsara Vilasa          500.0          4329.0            2.05   
Sobha City                             NaN             NaN             NaN   
Signature Global City 93               NaN           

In [ ]:
df_final_refined_v2['building type_Land'] = df_final_refined_v2['building type_Land'].replace({'':'Land'})

In [ ]:
df_final_refined_v2.head()

building type_1 BHK  area low 1 BHK  \
PropertyName                                                      
Smartworld One DXP                         None             NaN   
M3M Crown                                  None             NaN   
Adani Brahma Samsara Vilasa                None             NaN   
Sobha City                                 None             NaN   
Signature Global City 93                   None             NaN   

                             area high 1 BHK  price low 1 BHK  \
PropertyName                                                    
Smartworld One DXP                       NaN              NaN   
M3M Crown                                NaN              NaN   
Adani Brahma Samsara Vilasa              NaN              NaN   
Sobha City                               NaN              NaN   
Signature Global City 93                 NaN              NaN   

                             price high 1 BHK building type_2 BHK  \
PropertyName                                                        
Smartworld One DXP                        NaN           Apartment   
M3M Crown                                 NaN                None   
Adani Brahma Samsara Vilasa               NaN                None   
Sobha City                                NaN           Apartment   
Signature Global City 93                  NaN   Independent Floor   

                             area low 2 BHK  area high 2 BHK  price low 2 BHK  \
PropertyName                                                                    
Smartworld One DXP                   1370.0           1370.0           2.0000   
M3M Crown                               NaN              NaN              NaN   
Adani Brahma Samsara Vilasa             NaN              NaN              NaN   
Sobha City                           1381.0           1692.0           1.5500   
Signature Global City 93              981.0           1118.0           0.9301   

                             price high 2 BHK  ... building type_1 RK  \
PropertyName                                   ...                      
Smartworld One DXP                       2.40  ...               None   
M3M Crown                                 NaN  ...               None   
Adani Brahma Samsara Vilasa               NaN  ...               None   
Sobha City                               3.21  ...               None   
Signature Global City 93                 1.06  ...               None   

                             area low 1 RK  area high 1 RK  price low 1 RK  \
PropertyName                                                                 
Smartworld One DXP                     NaN             NaN             NaN   
M3M Crown                              NaN             NaN             NaN   
Adani Brahma Samsara Vilasa            NaN             NaN             NaN   
Sobha City                             NaN             NaN             NaN   
Signature Global City 93               NaN             NaN             NaN   

                             price high 1 RK building type_Land  \
PropertyName                                                      
Smartworld One DXP                       NaN               None   
M3M Crown                                NaN               None   
Adani Brahma Samsara Vilasa              NaN               Land   
Sobha City                               NaN               None   
Signature Global City 93                 NaN               None   

                             area low Land  area high Land  price low Land  \
PropertyName                                                                 
Smartworld One DXP                     NaN             NaN             NaN   
M3M Crown                              NaN             NaN             NaN   
Adani Brahma Samsara Vilasa          500.0          4329.0            2.05   
Sobha City                             NaN             NaN             NaN   
Signature Global City 93               NaN           

In [ ]:
categorical_columns = df_final_refined_v2.select_dtypes(include=['object']).columns.tolist()
categorical_columns

['building type_1 BHK',
 'building type_2 BHK',
 'building type_3 BHK',
 'building type_4 BHK',
 'building type_5 BHK',
 'building type_6 BHK',
 'building type_1 RK',
 'building type_Land']

In [ ]:
ohe_df = pd.get_dummies(df_final_refined_v2, columns=categorical_columns, drop_first=True)

In [ ]:
ohe_df.fillna(0,inplace=True)

In [ ]:
ohe_df.head()

area low 1 BHK  area high 1 BHK  price low 1 BHK  \
PropertyName                                                                    
Smartworld One DXP                      0.0              0.0              0.0   
M3M Crown                               0.0              0.0              0.0   
Adani Brahma Samsara Vilasa             0.0              0.0              0.0   
Sobha City                              0.0              0.0              0.0   
Signature Global City 93                0.0              0.0              0.0   

                             price high 1 BHK  area low 2 BHK  \
PropertyName                                                    
Smartworld One DXP                        0.0          1370.0   
M3M Crown                                 0.0             0.0   
Adani Brahma Samsara Vilasa               0.0             0.0   
Sobha City                                0.0          1381.0   
Signature Global City 93                  0.0           981.0   

                             area high 2 BHK  price low 2 BHK  \
PropertyName                                                    
Smartworld One DXP                    1370.0           2.0000   
M3M Crown                                0.0           0.0000   
Adani Brahma Samsara Vilasa              0.0           0.0000   
Sobha City                            1692.0           1.5500   
Signature Global City 93              1118.0           0.9301   

                             price high 2 BHK  area low 3 BHK  \
PropertyName                                                    
Smartworld One DXP                       2.40          1850.0   
M3M Crown                                0.00          1605.0   
Adani Brahma Samsara Vilasa              0.00          1800.0   
Sobha City                               3.21          1711.0   
Signature Global City 93                 1.06          1235.0   

                             area high 3 BHK  ...  \
PropertyName                                  ...   
Smartworld One DXP                    2050.0  ...   
M3M Crown                             2170.0  ...   
Adani Brahma Samsara Vilasa           3150.0  ...   
Sobha City                            2343.0  ...   
Signature Global City 93              1530.0  ...   

                             building type_2 BHK_Independent Floor  \
PropertyName                                                         
Smartworld One DXP                                           False   
M3M Crown                                                    False   
Adani Brahma Samsara Vilasa                                  False   
Sobha City                                                   False   
Signature Global City 93                                      True   

                             building type_2 BHK_Service Apartment  \
PropertyName                                                         
Smartworld One DXP                                           False   
M3M Crown                                                    False   
Adani Brahma Samsara Vilasa                                  False   
Sobha City                                                   False   
Signature Global City 93                                     False   

                             building type_3 BHK_Independent Floor  \
PropertyName                                                         
Smartworld One DXP                                           False   
M3M Crown                                                    False   
Adani Brahma Samsara Vilasa                                   True   
Sobha City                                                   False   
Signature Global City 93                                      True   

                             building type_3 BHK_Service Apartment  \
PropertyName                                                         
Smartworld One DXP                                           False   
M3M Crown                             

In [ ]:
from sklearn.preprocessing import StandardScaler

# Initialize the scaler
scaler = StandardScaler()

# Apply the scaler to the entire dataframe
ohe_df_normalized = pd.DataFrame(scaler.fit_transform(ohe_df), columns=ohe_df.columns, index=ohe_df.index)



In [ ]:
ohe_df_normalized.head()

area low 1 BHK  area high 1 BHK  price low 1 BHK  \
PropertyName                                                                    
Smartworld One DXP                -0.252266        -0.169584        -0.105197   
M3M Crown                         -0.252266        -0.169584        -0.105197   
Adani Brahma Samsara Vilasa       -0.252266        -0.169584        -0.105197   
Sobha City                        -0.252266        -0.169584        -0.105197   
Signature Global City 93          -0.252266        -0.169584        -0.105197   

                             price high 1 BHK  area low 2 BHK  \
PropertyName                                                    
Smartworld One DXP                  -0.082332        1.223499   
M3M Crown                           -0.082332       -0.893541   
Adani Brahma Samsara Vilasa         -0.082332       -0.893541   
Sobha City                          -0.082332        1.240497   
Signature Global City 93            -0.082332        0.622383   

                             area high 2 BHK  price low 2 BHK  \
PropertyName                                                    
Smartworld One DXP                  1.020101        -0.173712   
M3M Crown                          -0.896660        -0.283546   
Adani Brahma Samsara Vilasa        -0.896660        -0.283546   
Sobha City                          1.470610        -0.198425   
Signature Global City 93            0.667529        -0.232468   

                             price high 2 BHK  area low 3 BHK  \
PropertyName                                                    
Smartworld One DXP                   1.158423        0.553787   
M3M Crown                           -0.387986        0.293086   
Adani Brahma Samsara Vilasa         -0.387986        0.500583   
Sobha City                           1.680336        0.405879   
Signature Global City 93             0.295011       -0.100626   

                             area high 3 BHK  ...  \
PropertyName                                  ...   
Smartworld One DXP                  0.370864  ...   
M3M Crown                           0.472749  ...   
Adani Brahma Samsara Vilasa         1.304803  ...   
Sobha City                          0.619632  ...   
Signature Global City 93           -0.070634  ...   

                             building type_2 BHK_Independent Floor  \
PropertyName                                                         
Smartworld One DXP                                       -0.289310   
M3M Crown                                                -0.289310   
Adani Brahma Samsara Vilasa                              -0.289310   
Sobha City                                               -0.289310   
Signature Global City 93                                  3.456497   

                             building type_2 BHK_Service Apartment  \
PropertyName                                                         
Smartworld One DXP                                       -0.063888   
M3M Crown                                                -0.063888   
Adani Brahma Samsara Vilasa                              -0.063888   
Sobha City                                               -0.063888   
Signature Global City 93                                 -0.063888   

                             building type_3 BHK_Independent Floor  \
PropertyName                                                         
Smartworld One DXP                                       -0.372678   
M3M Crown                                                -0.372678   
Adani Brahma Samsara Vilasa                               2.683282   
Sobha City                                               -0.372678   
Signature Global City 93                                  2.683282   

                             building type_3 BHK_Service Apartment  \
PropertyName                                                         
Smartworld One DXP                                       -0.063888   
M3M Crown                             

In [ ]:
ohe_df_normalized.shape

(246, 43)

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

# Compute the cosine similarity matrix
cosine_sim2 = cosine_similarity(ohe_df_normalized)

In [ ]:
cosine_sim2.shape

(246, 246)

In [ ]:
def recommend_properties_with_scores(property_name, top_n=247):

    # Get the similarity scores for the property using its name as the index
    sim_scores = list(enumerate(cosine_sim2[ohe_df_normalized.index.get_loc(property_name)]))

    # Sort properties based on the similarity scores
    sorted_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the indices and scores of the top_n most similar properties
    top_indices = [i[0] for i in sorted_scores[1:top_n+1]]
    top_scores = [i[1] for i in sorted_scores[1:top_n+1]]

    # Retrieve the names of the top properties using the indices
    top_properties = ohe_df_normalized.index[top_indices].tolist()

    # Create a dataframe with the results
    recommendations_df = pd.DataFrame({
        'PropertyName': top_properties,
        'SimilarityScore': top_scores
    })

    return recommendations_df

# Test the recommender function using a property name
recommend_properties_with_scores('M3M Golf Hills')

PropertyName  SimilarityScore
0    AIPL The Peaceful Homes         0.955462
1         Smartworld One DXP         0.954670
2             Unitech Escape         0.953092
3                M3M Capital         0.951156
4                 BPTP Terra         0.943128
..                       ...              ...
240              Golden Park        -0.522391
241      Satya Merano Greens        -0.523660
242       ROF Normanton Park        -0.525129
243          BPTP Green Oaks        -0.525286
244           JMS The Nation        -0.526287

[245 rows x 2 columns]

## LocationAdvantages

In [ ]:
df[['PropertyName','LocationAdvantages']]['LocationAdvantages'][0]

"{'Bajghera Road': '800 Meter', 'Palam Vihar Halt': '2.5 KM', 'DPSG Palam Vihar': '3.1 KM', 'Park Hospital': '3.1 KM', 'Gurgaon Railway Station': '4.9 KM', 'The NorthCap University': '5.4 KM', 'Dwarka Expy': '1.2 KM', 'Hyatt Place Gurgaon Udyog Vihar': '7.7 KM', 'Dwarka Sector 21, Metro Station': '7.2 KM', 'Pacific D21 Mall': '7.4 KM', 'Indira Gandhi International Airport': '14.7 KM', 'Hamoni Golf Camp': '6.2 KM', 'Fun N Food Waterpark': '8.8 KM', 'Accenture DDC5': '9 KM'}"

In [ ]:
def distance_to_meters(distance_str):
    try:
        if 'Km' in distance_str or 'KM' in distance_str:
            return float(distance_str.split()[0]) * 1000
        elif 'Meter' in distance_str or 'meter' in distance_str:
            return float(distance_str.split()[0])
        else:
            return None
    except:
        return None

In [ ]:
import ast
# Extract distances for each location
location_matrix = {}
for index, row in df.iterrows():
    distances = {}
    for location, distance in ast.literal_eval(row['LocationAdvantages']).items():
        distances[location] = distance_to_meters(distance)
    location_matrix[index] = distances

# Convert the dictionary to a dataframe
location_df = pd.DataFrame.from_dict(location_matrix, orient='index')

# Display the first few rows
location_df.head()

Bajghera Road  Palam Vihar Halt  DPSG Palam Vihar  Park Hospital  \
0           800.0            2500.0            3100.0         3100.0   
25          550.0               NaN               NaN            NaN   
37         5300.0               NaN               NaN            NaN   
69         1500.0               NaN               NaN            NaN   
9             NaN               NaN               NaN         5500.0   

    Gurgaon Railway Station  The NorthCap University  Dwarka Expy  \
0                    4900.0                   5400.0       1200.0   
25                      NaN                   6700.0       3800.0   
37                   2500.0                   8800.0          NaN   
69                   6500.0                   6700.0       5100.0   
9                       NaN                      NaN          NaN   

    Hyatt Place Gurgaon Udyog Vihar  Dwarka Sector 21, Metro Station  \
0                            7700.0                           7200.0   
25                              NaN                              NaN   
37                              NaN                              NaN   
69                              NaN                              NaN   
9                               NaN                              NaN   

    Pacific D21 Mall  ...  MCC Cricket Ground Dhankot  \
0             7400.0  ...                         NaN   
25            7500.0  ...                         NaN   
37               NaN  ...                         NaN   
69            8200.0  ...                         NaN   
9                NaN  ...                         NaN   

    The Shri Ram School Aravali  Taj City Centre Gurugram  \
0                           NaN                       NaN   
25                          NaN                       NaN   
37                          NaN                       NaN   
69                          NaN                       NaN   
9                           NaN                       NaN   

    Minda Industries  Corporate Office  Rampura Flyover, Naurangpur Rd  \
0                                  NaN                             NaN   
25                                 NaN                             NaN   
37                                 NaN                             NaN   
69                                 NaN                             NaN   
9                                  NaN                             NaN   

    Manesar toll plaza - Kherki Daula  Imt Manesar, Gurugram  Holiday Inn  \
0                                 NaN                    NaN          NaN   
25                                NaN                    NaN          NaN   
37                                NaN                    NaN          NaN   
69                                NaN                    NaN          NaN   
9                                 NaN                    NaN          NaN   

    Sector 84 Road  Skyview Corporate Park  
0              NaN                     NaN  
25             NaN                     NaN  
37             NaN                     NaN  
69             NaN                     NaN  
9              NaN                     NaN  

[5 rows x 1070 columns]

### Here we got 2 probelems
* same Landmark are having different spelling:
* How to remove NaN value (we replaced this value with highest value)

In [ ]:
location_df.columns[10:50]

Index(['Indira Gandhi International Airport', 'Hamoni Golf Camp',
       'Fun N Food Waterpark', 'Accenture DDC5', 'DPSG Palam Vihar Gurugram',
       'Park Hospital, Palam Vihar', 'Palam Vihar Halt Railway Station',
       'Dwarka Sector 21 Metro Station', 'Dwarka Expressway',
       'Fun N Food Water Park', 'Tau DeviLal Sports Complex', 'Hyatt Place',
       'Altrade Business Centre', 'AIPL Business Club Sector 62',
       'Heritage Xperiential Learning School', 'CK Birla Hospital',
       'Paras Trinity Mall Sector 63', 'Rapid Metro Station Sector 56',
       'De Adventure Park', 'Golf Course Ext Rd',
       'DoubleTree by Hilton Hotel Gurgaon',
       'KIIT College of Engineering Sohna Road', 'Mehrauli-Gurgaon Road',
       'Nirvana Rd', 'TERI Golf Course', 'The Shikshiyan School', 'WTC Plaza',
       'Luxus Haritma Resort', 'BSF Golf Course', 'Rions Hospital', 'Gurgaon',
       'Dwarka Sector 21', 'Nehru Stadium', 'Fun N Food WaterPark',
       'IGI Airport', 'Vasant Kunj', 'Prana

In [ ]:
location_df.index = df.PropertyName

In [ ]:
location_df.head()

Bajghera Road  Palam Vihar Halt  \
PropertyName                                                   
Smartworld One DXP                   800.0            2500.0   
M3M Crown                            550.0               NaN   
Adani Brahma Samsara Vilasa         5300.0               NaN   
Sobha City                          1500.0               NaN   
Signature Global City 93               NaN               NaN   

                             DPSG Palam Vihar  Park Hospital  \
PropertyName                                                   
Smartworld One DXP                     3100.0         3100.0   
M3M Crown                                 NaN            NaN   
Adani Brahma Samsara Vilasa               NaN            NaN   
Sobha City                                NaN            NaN   
Signature Global City 93                  NaN         5500.0   

                             Gurgaon Railway Station  The NorthCap University  \
PropertyName                                                                    
Smartworld One DXP                            4900.0                   5400.0   
M3M Crown                                        NaN                   6700.0   
Adani Brahma Samsara Vilasa                   2500.0                   8800.0   
Sobha City                                    6500.0                   6700.0   
Signature Global City 93                         NaN                      NaN   

                             Dwarka Expy  Hyatt Place Gurgaon Udyog Vihar  \
PropertyName                                                                
Smartworld One DXP                1200.0                           7700.0   
M3M Crown                         3800.0                              NaN   
Adani Brahma Samsara Vilasa          NaN                              NaN   
Sobha City                        5100.0                              NaN   
Signature Global City 93             NaN                              NaN   

                             Dwarka Sector 21, Metro Station  \
PropertyName                                                   
Smartworld One DXP                                    7200.0   
M3M Crown                                                NaN   
Adani Brahma Samsara Vilasa                              NaN   
Sobha City                                               NaN   
Signature Global City 93                                 NaN   

                             Pacific D21 Mall  ...  \
PropertyName                                   ...   
Smartworld One DXP                     7400.0  ...   
M3M Crown                              7500.0  ...   
Adani Brahma Samsara Vilasa               NaN  ...   
Sobha City                             8200.0  ...   
Signature Global City 93                  NaN  ...   

                             MCC Cricket Ground Dhankot  \
PropertyName                                              
Smartworld One DXP                                  NaN   
M3M Crown                                           NaN   
Adani Brahma Samsara Vilasa                         NaN   
Sobha City                                          NaN   
Signature Global City 93                            NaN   

                             The Shri Ram School Aravali  \
PropertyName                                               
Smartworld One DXP                                   NaN   
M3M Crown                                            NaN   
Adani Brahma Samsara Vilasa                          NaN   
Sobha City                                           NaN   
Signature Global City 93                             NaN   

                             Taj City Centre Gurugram  \
PropertyName                                            
Smartworld One DXP                                NaN   
M3M Crown                                         NaN   
Adani Brahma Samsara Vilasa                       NaN   
Sobha City                                        NaN   
Signature Global C

In [ ]:
location_df.fillna(54000,inplace=True)

In [ ]:
location_df.head()

Bajghera Road  Palam Vihar Halt  \
PropertyName                                                   
Smartworld One DXP                   800.0            2500.0   
M3M Crown                            550.0           54000.0   
Adani Brahma Samsara Vilasa         5300.0           54000.0   
Sobha City                          1500.0           54000.0   
Signature Global City 93           54000.0           54000.0   

                             DPSG Palam Vihar  Park Hospital  \
PropertyName                                                   
Smartworld One DXP                     3100.0         3100.0   
M3M Crown                             54000.0        54000.0   
Adani Brahma Samsara Vilasa           54000.0        54000.0   
Sobha City                            54000.0        54000.0   
Signature Global City 93              54000.0         5500.0   

                             Gurgaon Railway Station  The NorthCap University  \
PropertyName                                                                    
Smartworld One DXP                            4900.0                   5400.0   
M3M Crown                                    54000.0                   6700.0   
Adani Brahma Samsara Vilasa                   2500.0                   8800.0   
Sobha City                                    6500.0                   6700.0   
Signature Global City 93                     54000.0                  54000.0   

                             Dwarka Expy  Hyatt Place Gurgaon Udyog Vihar  \
PropertyName                                                                
Smartworld One DXP                1200.0                           7700.0   
M3M Crown                         3800.0                          54000.0   
Adani Brahma Samsara Vilasa      54000.0                          54000.0   
Sobha City                        5100.0                          54000.0   
Signature Global City 93         54000.0                          54000.0   

                             Dwarka Sector 21, Metro Station  \
PropertyName                                                   
Smartworld One DXP                                    7200.0   
M3M Crown                                            54000.0   
Adani Brahma Samsara Vilasa                          54000.0   
Sobha City                                           54000.0   
Signature Global City 93                             54000.0   

                             Pacific D21 Mall  ...  \
PropertyName                                   ...   
Smartworld One DXP                     7400.0  ...   
M3M Crown                              7500.0  ...   
Adani Brahma Samsara Vilasa           54000.0  ...   
Sobha City                             8200.0  ...   
Signature Global City 93              54000.0  ...   

                             MCC Cricket Ground Dhankot  \
PropertyName                                              
Smartworld One DXP                              54000.0   
M3M Crown                                       54000.0   
Adani Brahma Samsara Vilasa                     54000.0   
Sobha City                                      54000.0   
Signature Global City 93                        54000.0   

                             The Shri Ram School Aravali  \
PropertyName                                               
Smartworld One DXP                               54000.0   
M3M Crown                                        54000.0   
Adani Brahma Samsara Vilasa                      54000.0   
Sobha City                                       54000.0   
Signature Global City 93                         54000.0   

                             Taj City Centre Gurugram  \
PropertyName                                            
Smartworld One DXP                            54000.0   
M3M Crown                                     54000.0   
Adani Brahma Samsara Vilasa                   54000.0   
Sobha City                                    54000.0   
Signature Global C

In [ ]:
from sklearn.preprocessing import StandardScaler
# Initialize the scaler
scaler = StandardScaler()

# Apply the scaler to the entire dataframe
location_df_normalized = pd.DataFrame(scaler.fit_transform(location_df), columns=location_df.columns, index=location_df.index)

In [ ]:
cosine_sim3 = cosine_similarity(location_df_normalized)

In [ ]:
cosine_sim1.shape

(246, 246)

In [ ]:
def recommend_properties_with_scores(property_name, top_n=247):

    cosine_sim_matrix = 30*cosine_sim1 + 20*cosine_sim2 + 8*cosine_sim3
    # cosine_sim_matrix = cosine_sim3

    # Get the similarity scores for the property using its name as the index
    sim_scores = list(enumerate(cosine_sim_matrix[location_df_normalized.index.get_loc(property_name)]))

    # Sort properties based on the similarity scores
    sorted_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the indices and scores of the top_n most similar properties
    top_indices = [i[0] for i in sorted_scores[1:top_n+1]]
    top_scores = [i[1] for i in sorted_scores[1:top_n+1]]

    # Retrieve the names of the top properties using the indices
    top_properties = location_df_normalized.index[top_indices].tolist()

    # Create a dataframe with the results
    recommendations_df = pd.DataFrame({
        'PropertyName': top_properties,
        'SimilarityScore': top_scores
    })

    return recommendations_df

# Test the recommender function using a property name
recommend_properties_with_scores('Ireo Victory Valley')

PropertyName  SimilarityScore
0    Pioneer Urban Presidia        28.021460
1        Ambience Creacions        27.787913
2             DLF The Crest        24.205986
3             Pioneer Araya        23.415308
4    Silverglades The Melia        21.007840
..                      ...              ...
240          JMS The Nation       -14.735656
241     Shree Vardhman City       -14.888274
242          JMS Prime Land       -15.082229
243       Vatika Aspiration       -15.099679
244           Corona Greens       -15.673834

[245 rows x 2 columns]

### Here we have flexibility to assign the weightage to each recommender sysytem
* cosine_sim1 ----> Facilities
* cosine_sim2 ----> Pricing
* cosine_sim3 ----> Locations

We will send the Apartment's name to recommender system to work

#### we have tested the recommender system by sending **DLF The Camellias** property name which is costlier apartment, should also return the costly by using price recommender system, similarly for others

In [ ]:
(3*cosine_sim3 + 5*cosine_sim2 + 6*cosine_sim1).shape

(246, 246)

In [ ]:
df[df['PropertyName'].str.contains("DLF The Camellias")]

In [ ]:
location_df

Bajghera Road  Palam Vihar Halt  \
PropertyName                                                   
Smartworld One DXP                   800.0            2500.0   
M3M Crown                            550.0           54000.0   
Adani Brahma Samsara Vilasa         5300.0           54000.0   
Sobha City                          1500.0           54000.0   
Signature Global City 93           54000.0           54000.0   
...                                    ...               ...   
DLF Princeton Estate               54000.0           54000.0   
Pyramid Urban Homes 2              54000.0           54000.0   
Satya The Hermitage                54000.0           54000.0   
BPTP Spacio                        54000.0           54000.0   
SS The Coralwood                   54000.0           54000.0   

                             DPSG Palam Vihar  Park Hospital  \
PropertyName                                                   
Smartworld One DXP                     3100.0         3100.0   
M3M Crown                             54000.0        54000.0   
Adani Brahma Samsara Vilasa           54000.0        54000.0   
Sobha City                            54000.0        54000.0   
Signature Global City 93              54000.0         5500.0   
...                                       ...            ...   
DLF Princeton Estate                  54000.0        54000.0   
Pyramid Urban Homes 2                 54000.0        54000.0   
Satya The Hermitage                   54000.0        54000.0   
BPTP Spacio                           54000.0        54000.0   
SS The Coralwood                      54000.0        54000.0   

                             Gurgaon Railway Station  The NorthCap University  \
PropertyName                                                                    
Smartworld One DXP                            4900.0                   5400.0   
M3M Crown                                    54000.0                   6700.0   
Adani Brahma Samsara Vilasa                   2500.0                   8800.0   
Sobha City                                    6500.0                   6700.0   
Signature Global City 93                     54000.0                  54000.0   
...                                              ...                      ...   
DLF Princeton Estate                         54000.0                  54000.0   
Pyramid Urban Homes 2                        54000.0                  54000.0   
Satya The Hermitage                          54000.0                  54000.0   
BPTP Spacio                                  54000.0                  54000.0   
SS The Coralwood                             54000.0                  54000.0   

                             Dwarka Expy  Hyatt Place Gurgaon Udyog Vihar  \
PropertyName                                                                
Smartworld One DXP                1200.0                           7700.0   
M3M Crown                         3800.0                          54000.0   
Adani Brahma Samsara Vilasa      54000.0                          54000.0   
Sobha City                        5100.0                          54000.0   
Signature Global City 93         54000.0                          54000.0   
...                                  ...                              ...   
DLF Princeton Estate             54000.0                          54000.0   
Pyramid Urban Homes 2            54000.0                          54000.0   
Satya The Hermitage              54000.0                          54000.0   
BPTP Spacio                      54000.0                          54000.0   
SS The Coralwood                 54000.0                          54000.0   

                             Dwarka Sector 21, Metro Station  \
PropertyName                                                   
Smartworld One DXP                                    7200.0   
M3M Crown                                            54000.0   
Adani Brahma Samsara Vilasa                        

In [ ]:
import pickle
pickle.dump(location_df,open('location_distance.pkl','wb'))

In [ ]:
pickle.dump(cosine_sim1,open('cosine_sim1.pkl','wb'))
pickle.dump(cosine_sim2,open('cosine_sim2.pkl','wb'))
pickle.dump(cosine_sim3,open('cosine_sim3.pkl','wb'))

In [ ]:
df = df[['PropertyName', 'Link']]

In [ ]:
pickle.dump(df,open('main_df.pkl','wb'))

In [ ]:
df[df['PropertyName']== 'Smartworld One DXP']

PropertyName                                               Link
0  Smartworld One DXP  https://www.99acres.com/smartworld-one-dxp-sec...